File: python/getexp-core/getexp_core/storage.py
"""Storage ports (hexagonal / ports & adapters).

App code depends on these protocols, never on a database driver or cloud SDK
directly. Concrete adapters (Postgres, S3, Azure Blob, ...) live in separate
optional packages and are selected at boot by configuration (a connection URL).
That is what makes storage plug-and-play across infra: swapping a vendor is a
config change, not a code change.

This module ships the contracts plus in-memory adapters for tests and
single-instance/dev use. Mirrors ``@getexp/core``'s storage module. See
docs/adr/0004-storage-ports-and-adapters.md.
"""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass, field
from typing import Generic, Protocol, TypeVar
from urllib.parse import quote

from getexp_core.errors import NotFoundError

T = TypeVar("T")
Id = TypeVar("Id")
# Id appears only in parameter (input) positions on the protocol, so it must be
# declared contravariant; the concrete InMemoryRepository uses an invariant Id.
IdContra = TypeVar("IdContra", contravariant=True)

# ---------------------------------------------------------------------------
# Repository — the common-case CRUD contract over a collection of entities.
# ---------------------------------------------------------------------------


class Repository(Protocol[T, IdContra]):
    async def get(self, id: IdContra) -> T | None: ...
    async def list(self) -> list[T]: ...
    async def create(self, entity: T) -> T: ...
    async def update(self, id: IdContra, patch: dict[str, object]) -> T: ...
    async def delete(self, id: IdContra) -> None: ...


class InMemoryRepository(Generic[T, Id]):
    """In-memory Repository for tests/dev. ``id_of`` extracts an entity's id."""

    def __init__(self, id_of: Callable[[T], Id]) -> None:
        self._id_of = id_of
        self._items: dict[Id, T] = {}

    async def get(self, id: Id) -> T | None:
        return self._items.get(id)

    async def list(self) -> list[T]:
        return list(self._items.values())

    async def create(self, entity: T) -> T:
        self._items[self._id_of(entity)] = entity
        return entity

    async def update(self, id: Id, patch: dict[str, object]) -> T:
        existing = self._items.get(id)
        if existing is None:
            raise NotFoundError(f"Entity {id} not found")
        for key, value in patch.items():
            setattr(existing, key, value)
        return existing

    async def delete(self, id: Id) -> None:
        self._items.pop(id, None)


# ---------------------------------------------------------------------------
# BlobStore — object/blob storage behind one contract. Adapters: S3-compatible
# (AWS S3 / GCS interop / Cloudflare R2 / MinIO) and Azure Blob Storage.
# ---------------------------------------------------------------------------


@dataclass(frozen=True, slots=True)
class PutOptions:
    content_type: str | None = None
    metadata: dict[str, str] = field(default_factory=dict)


@dataclass(frozen=True, slots=True)
class PresignOptions:
    expires_in_seconds: int = 3600
    operation: str = "get"


class BlobStore(Protocol):
    async def put(self, key: str, data: bytes, options: PutOptions | None = None) -> None: ...
    async def get(self, key: str) -> bytes: ...
    async def exists(self, key: str) -> bool: ...
    async def delete(self, key: str) -> None: ...
    async def presigned_url(self, key: str, options: PresignOptions | None = None) -> str: ...


class InMemoryBlobStore:
    def __init__(self) -> None:
        self._blobs: dict[str, bytes] = {}

    async def put(self, key: str, data: bytes, options: PutOptions | None = None) -> None:
        self._blobs[key] = data

    async def get(self, key: str) -> bytes:
        blob = self._blobs.get(key)
        if blob is None:
            raise NotFoundError(f"Blob {key} not found")
        return blob

    async def exists(self, key: str) -> bool:
        return key in self._blobs

    async def delete(self, key: str) -> None:
        self._blobs.pop(key, None)

    async def presigned_url(self, key: str, options: PresignOptions | None = None) -> str:
        opts = options or PresignOptions()
        return f"memory://blob/{quote(key)}?op={opts.operation}&expires={opts.expires_in_seconds}"


# ---------------------------------------------------------------------------
# HealthCheck — storage adapters report readiness so /ready reflects reality.
# ---------------------------------------------------------------------------


class HealthCheck(Protocol):
    name: str

    async def ping(self) -> bool: ...


@dataclass(frozen=True, slots=True)
class HealthReport:
    healthy: bool
    checks: dict[str, bool]


async def check_health(checks: list[HealthCheck]) -> HealthReport:
    """Ping every dependency; a failed ping (or raise) marks it unhealthy."""
    results: dict[str, bool] = {}
    for check in checks:
        try:
            results[check.name] = await check.ping()
        except Exception:
            results[check.name] = False
    return HealthReport(healthy=all(results.values()), checks=results)

File: packages/core/src/storage.ts
/**
 * Storage ports (hexagonal / ports & adapters).
 *
 * App code depends on these interfaces, never on a database driver or cloud
 * SDK directly. Concrete adapters (Postgres, S3, Azure Blob, …) live in
 * separate optional packages and are selected at boot by configuration (a
 * connection URL). That is what makes storage plug-and-play across infra:
 * swapping a vendor is a config change, not a code change.
 *
 * This module ships the contracts plus in-memory adapters for tests and
 * single-instance/dev use. See docs/adr/0004-storage-ports-and-adapters.md.
 */
import { NotFoundError } from './errors.ts';

// ---------------------------------------------------------------------------
// Repository — the common-case CRUD contract over a collection of entities.
// Complex queries may drop to the underlying ORM; repositories keep the 80%
// case portable and trivially testable with InMemoryRepository.
// ---------------------------------------------------------------------------

export interface Repository<T, Id = string> {
  get(id: Id): Promise<T | null>;
  list(): Promise<T[]>;
  create(entity: T): Promise<T>;
  update(id: Id, patch: Partial<T>): Promise<T>;
  delete(id: Id): Promise<void>;
}

export class InMemoryRepository<T, Id = string> implements Repository<T, Id> {
  private readonly items = new Map<Id, T>();

  constructor(private readonly idOf: (entity: T) => Id) {}

  async get(id: Id): Promise<T | null> {
    return this.items.get(id) ?? null;
  }

  async list(): Promise<T[]> {
    return [...this.items.values()];
  }

  async create(entity: T): Promise<T> {
    this.items.set(this.idOf(entity), entity);
    return entity;
  }

  async update(id: Id, patch: Partial<T>): Promise<T> {
    const existing = this.items.get(id);
    if (existing === undefined) throw new NotFoundError(`Entity ${String(id)} not found`);
    const updated = { ...existing, ...patch };
    this.items.set(id, updated);
    return updated;
  }

  async delete(id: Id): Promise<void> {
    this.items.delete(id);
  }
}

// ---------------------------------------------------------------------------
// BlobStore — object/blob storage behind one contract. Adapters: S3-compatible
// (AWS S3 / GCS interop / Cloudflare R2 / MinIO) and Azure Blob Storage.
// ---------------------------------------------------------------------------

export interface PutOptions {
  contentType?: string;
  metadata?: Record<string, string>;
}

export interface PresignOptions {
  /** URL validity in seconds. Default 3600. */
  expiresInSeconds?: number;
  /** Whether the URL is for download or upload. Default 'get'. */
  operation?: 'get' | 'put';
}

export interface BlobStore {
  put(key: string, data: Uint8Array, options?: PutOptions): Promise<void>;
  get(key: string): Promise<Uint8Array>;
  exists(key: string): Promise<boolean>;
  delete(key: string): Promise<void>;
  /** A time-limited URL clients can use directly, bypassing the service. */
  presignedUrl(key: string, options?: PresignOptions): Promise<string>;
}

export class InMemoryBlobStore implements BlobStore {
  private readonly blobs = new Map<string, { data: Uint8Array; options?: PutOptions }>();

  async put(key: string, data: Uint8Array, options?: PutOptions): Promise<void> {
    this.blobs.set(key, options ? { data, options } : { data });
  }

  async get(key: string): Promise<Uint8Array> {
    const blob = this.blobs.get(key);
    if (!blob) throw new NotFoundError(`Blob ${key} not found`);
    return blob.data;
  }

  async exists(key: string): Promise<boolean> {
    return this.blobs.has(key);
  }

  async delete(key: string): Promise<void> {
    this.blobs.delete(key);
  }

  async presignedUrl(key: string, options: PresignOptions = {}): Promise<string> {
    const expires = options.expiresInSeconds ?? 3600;
    return `memory://blob/${encodeURIComponent(key)}?op=${options.operation ?? 'get'}&expires=${expires}`;
  }
}

// ---------------------------------------------------------------------------
// HealthCheck — storage adapters report readiness so /ready reflects reality.
// ---------------------------------------------------------------------------

export interface HealthCheck {
  readonly name: string;
  ping(): Promise<boolean>;
}

export interface HealthReport {
  healthy: boolean;
  checks: Record<string, boolean>;
}

/** Ping every dependency; a failed ping (or throw) marks that check unhealthy. */
export async function checkHealth(checks: HealthCheck[]): Promise<HealthReport> {
  const entries = await Promise.all(
    checks.map(async (c) => {
      try {
        return [c.name, await c.ping()] as const;
      } catch {
        return [c.name, false] as const;
      }
    }),
  );
  const result: Record<string, boolean> = {};
  for (const [name, ok] of entries) result[name] = ok;
  return { healthy: entries.every(([, ok]) => ok), checks: result };
}
